lgc/disassembler/Disassembler.cpp

```diff
@@ -0,0 +1,506 @@
+      std::string comment;
+      raw_string_ostream commentStream(comment);
+      MCDisassembler::DecodeStatus status = MCDisassembler::Fail;
+      if (isCode && (offset & instAlignment - 1) == 0)
+        status = m_instDisassembler->getInstruction(inst, instSize, code, offset, commentStream);
+
+      if (status != MCDisassembler::Fail) {
+        // Got a disassemblable instruction.
+        // First output any non-disassemblable data up to this point.
+        if (lastOffset != offset) {
+          outputData(outputting, lastOffset, contents.slice(lastOffset, offset), nextReloc, relocs.end());
+          lastOffset = offset;
+        }
+        // Output reloc.
+        outputRelocs(outputting, offset, instSize, nextReloc, relocs.end());
+
+        if (outputting) {
+          // Output the binary encoding as a comment.
+          if (!comment.empty())
+            commentStream << " ";
+          if (status == MCDisassembler::SoftFail)
+            m_streamer->AddComment("Illegal instruction encoding ");
+          commentStream << format("%06x:", offset);
+          assert((instSize & (instAlignment - 1)) == 0);
+          for (size_t subOffset = 0; subOffset != instSize; ++subOffset) {
+            // This puts a space before each group of instAlignment bytes, and swizzles by
+            // instAlignment if little-endian. That has the effect of dumping words of size
+            // instAlignment.
+            if ((subOffset & (instAlignment - 1)) == 0)
+              commentStream << " ";
+            unsigned byte = code[subOffset];
+            if (endian() == support::little)
+              byte = code[subOffset ^ (instAlignment - 1)];
+            commentStream << format("%02x", byte);
+          }
+          // Output the instruction to the streamer.
+          if (!comment.empty())
+            m_streamer->AddComment(comment);
+          m_streamer->emitInstruction(inst, *m_subtargetInfo);
+        }
+        offset += instSize;
+        lastOffset = offset;
+        continue;
+      }
+      // No disassemblable instruction here. Try the next instruction unit.
+      offset = std::min(size_t(alignTo(offset + 1, instAlignment)), endOffset);
+    }
+
+    if (outputting)
+      break;
+
+    // If we created a symbolizer, get the referenced addresses from it and synthesize labels, avoiding
+    // duplicates. If there were no new referenced addresses, then we can do the final output in the next pass.
+    outputting = true;
+    if (symbolizerPtr) {
+      if (!symbolizerPtr->getReferencedAddresses().empty()) {
+        outputting = false;
+        std::vector<uint64_t> referencedAddresses;
+        referencedAddresses.insert(referencedAddresses.begin(), symbolizerPtr->getReferencedAddresses().begin(),
+                                   symbolizerPtr->getReferencedAddresses().end());
+        std::sort(referencedAddresses.begin(), referencedAddresses.end());
+        uint64_t lastLabel = ~0ULL;
```

Comment:
nit: `uint64_t lastLabel = ~uint64_t(0);` or `= kInvalidValue`? This is a bit pedantic, but I think `ULL` does not have to produce an `uint64_t` literal

Reply:
nit: Also, use llvm::sort above